src/libcore/slice/rotate.rs

```diff
@@ -2,88 +2,170 @@ use crate::cmp;
+        if left + right < 24 {
+            // Algorithm 1
+            // Microbenchmarks indicate that the average performance for random shifts is better all
+            // the way until about `left + right == 32`, but the worst case performance breaks even
+            // around 16. 24 was chosen as middle ground.
+            // Additionally, algorithm 1 is still faster than the others when `T` is huge.
+            let x = mid.sub(left);
+            // beginning of first round
+            let mut tmp: T = x.read();
+            let mut i = right;
+            // `gcd` can be found before hand by calculating `gcd(left + right, right)`,
+            // but it is faster to do one loop which calculates the gcd as a side effect, then
+            // doing the rest of the chunk
+            let mut gcd = right;
+            // benchmarks reveal that it is faster to swap temporaries all the way through instead
+            // of reading one temporary once, copying backwards, and then writing that temporary at
+            // the very end. This is probably due to the fact that swapping or replacing temporaries
+            // uses only one memory address in the loop instead of needing to manage two.
+            loop {
+                tmp = x.add(i).replace(tmp);
+                // instead of incrementing `i` and then checking if it is outside the bounds, we
+                // check if `i` will go outside the bounds on the next increment. This prevents
+                // any wrapping of pointers or `usize`.
+                if i >= left {
+                    i -= left;
+                    if i == 0 {
+                        // end of first round
+                        x.write(tmp);
+                        break;
+                    }
+                    // this conditional must be here if `left + right >= 15`
+                    if i < gcd {
+                        gcd = i;
+                    }
+                } else {
+                    i += right;
+                }
+            }
+            // finish the chunk with more rounds
+            for start in 1..gcd {
+                tmp = x.add(start).read();
+                i = start + right;
+                loop {
+                    tmp = x.add(i).replace(tmp);
+                    if i >= left {
+                        i -= left;
+                        if i == start {
+                            x.add(start).write(tmp);
+                            break;
+                        }
+                    } else {
+                        i += right;
+                    }
+                }
+            }
+            return;
+        // `T` is not a zero-sized type, so it's okay to divide by its size.
+        } else if cmp::min(left, right) <= mem::size_of::<BufType>() / mem::size_of::<T>() {
+            // Algorithm 2
+            let mut rawarray = MaybeUninit::<BufType>::uninit();
+            let buf = rawarray.as_mut_ptr() as *mut T;
```

Comment:
If `T` is more-aligned than `usize`, couldn't this cause under-aligned `memcpy`?  (See the comment on `RawArray` about alignment.  I guess you could also add a `[T;0]` field.)

Reply:
This is my first time messing with pointers, and since the `add` and `copy` operations work on `size_of<T>` automatically, I assumed that the code would work for anything except for ZSTs. I presume the original code avoided the alignment problems via the `[T;2]` field. What do you mean by a `[T;0]` field?

Reply:
Should I be using `align_of` instead of `size_of`, or am I misunderstanding something? I am not sure about the most performant and elegant way to do this.

Reply:
See the notes in <https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html#safety> -- the pointers that get passed into there need to be a multiple of the type's alignment.  `add` and `sub` preserve that property, but don't force it -- if the input pointer is misaligned the pointer that comes out is (probably) misaligned too.

The simplest fix is probably to just keep `BufType` as a `union`.  (The `[T;0]` thing is that you could probably do `MaybeUninit::<(BufType, [T;0])>::uninit()`, but that's kinda ugly.)

Reply:
I used the `[T; 0]` trick.